### PR TITLE
Streaming get with prefetch and Julia IO Stream API

### PR DESCRIPTION
**Pull request by andrebsguedes**

### CODE REVIEW

File: src/RustyObjectStore.jl [hunk 1]
```diff
@@ -555,4 +663,531 @@ function put_object(buffer::AbstractVector{UInt8}, path::String, conf::AbstractC
...
+        dest::Ptr{UInt8},
+        bytes_to_read::Culonglong,
+        bytes_to_read::Culonglong,
+        response_ref::Ref{ReadResponseFFI},
+        cond_handle::Ptr{Cvoid}
+    )::Cint
+
+    if result == 1
+        @error "failed to submit read_from_stream, runtime not started"
+        return nothing
```
[the remark this answers is not in the document]
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:949]: Done

File: src/RustyObjectStore.jl [hunk 2]
```diff
@@ -555,4 +663,531 @@ function put_object(buffer::AbstractVector{UInt8}, path::String, conf::AbstractC
...
+    ReadStreamResponseFFI() = new(-1, C_NULL, 0, C_NULL)
+end
+
+"""
+    ReadStream
+
+
+Opaque IO stream of object data.
+
+It is necessary to `finish!` the stream if it is not run to completion.
```
[the remark this answers is not in the document]
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:730]: And to clarify, the docs statement is correct, we should only need to close if we don't run it to completion

[the remark this answers is not in the document]
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:730]: This makes me remember we discussed adding a finalizer to ensure proper reclamation, let's see if we will have the time for them

File: src/RustyObjectStore.jl [hunk 3]
```diff
@@ -533,26 +680,526 @@ function put_object(buffer::AbstractVector{UInt8}, path::String, conf::AbstractC
...
+- `size_hint::Int`: (Optional) Expected size of the object (optimization for small objects).
+- `decompress::Option{String}`: (Optional) Compression algorithm to decode the response stream (supports gzip, deflate, zlib or zstd)
+
+# Returns
+- `stream::ReadStream`: The stream of object data chunks.
+
+# Throws
+- `GetException`: If the request fails for any reason.
+"""
+function get_object_stream(path::String, conf::AbstractConfig; size_hint::Int=0, decompress::String="")
```
Comment (Drvi) [inline src/RustyObjectStore.jl:868]: Could we call this `ReadStream` instead of `get_object_stream`?

File: src/RustyObjectStore.jl [hunk 4]
```diff
@@ -533,26 +680,526 @@ function put_object(buffer::AbstractVector{UInt8}, path::String, conf::AbstractC
...
+# Keyword
+- `compress::Option{String}`: (Optional) Compression algorithm to encode the stream (supports gzip, deflate, zlib or zstd)
+
+# Returns
+- `stream::WriteStream`: The stream where to write object data.
+
+# Throws
+- `PutException`: If the request fails for any reason.
+"""
+function put_object_stream(path::String, conf::AbstractConfig; compress::String="")
```
Comment (Drvi) [inline src/RustyObjectStore.jl:1031]: Can we call this `WriteStream` instead of `put_object_stream`?

**Review by Drvi — COMMENTED** [review]
Sorry for the late review! Left a couple of comments / suggestions which will hopefully be useful.

**Pull request merged**